pkg/reconciler/v1alpha1/pipelinerun/pipelinerun_test.go

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2018 The Knative Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pipelinerun
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/knative/build-pipeline/pkg/apis/pipeline/v1alpha1"
+	fakepipelineclientset "github.com/knative/build-pipeline/pkg/client/clientset/versioned/fake"
+	informers "github.com/knative/build-pipeline/pkg/client/informers/externalversions"
+	listers "github.com/knative/build-pipeline/pkg/client/listers/pipeline/v1alpha1"
+	"github.com/knative/build-pipeline/pkg/reconciler"
+	"github.com/knative/pkg/controller"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zaptest/observer"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	fakekubeclientset "k8s.io/client-go/kubernetes/fake"
+)
+
+func TestReconcile(t *testing.T) {
+	prs := []*v1alpha1.PipelineRun{{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-pipeline-run-success",
+			Namespace: "foo",
+		},
+		Spec: v1alpha1.PipelineRunSpec{
+			PipelineRef: v1alpha1.PipelineRef{
+				Name:       "test-pipeline",
+				APIVersion: "a1",
+			},
+		}}, {
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-pipeline-run-fetch-error",
+			Namespace: "foo",
+		}}, {
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "invalid-pipeline",
+			Namespace: "foo",
+		},
+		Spec: v1alpha1.PipelineRunSpec{
+			PipelineRef: v1alpha1.PipelineRef{
+				Name:       "pipeline-not-exist",
+				APIVersion: "a1",
+			},
+		}},
+	}
+
+	ps := []*v1alpha1.Pipeline{{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-pipeline",
+			Namespace: "foo",
+		}},
+	}
+	tcs := []struct {
+		name        string
+		pipelineRun string
+		shdErr      bool
+		log         string
+	}{
+		{"success", "foo/test-pipeline-run-success", false, ""},
+		{"pipeline-fetching-error-shd-error", "test-pipeline-run-fetch-error", true, ""},
+		{"invalid-pipeline-run-shd-succeed-with-logs", "foo/test-pipeline-run-doesnot-exist", false,
+			"pipeline run \"foo/test-pipeline-run-doesnot-exist\" in work queue no longer exists"},
+		{"invalid-pipeline-shd-succeed-with-logs", "foo/invalid-pipeline", false,
+			"\"foo/invalid-pipeline\" failed to Get Pipeline: \"foo/pipeline-not-exist\""},
+		{"invalid-pipeline-run-name-shd-succed-with-logs", "test/pipeline-fail/t", false,
+			"invalid resource key: test/pipeline-fail/t"},
+	}
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
```

Comment:
would be nice to use `https://github.com/google/go-cmp` to compare expected results vs actual results, for example compare the actual `PipelineRun` created. Something like [this](https://github.com/knative/build/blob/0f706d697bdc0def1acba236fbed193413935e91/pkg/builder/cluster/convert/convert_test.go#L134) maybe.
We can also add this later if you want